### PR TITLE
Openvino test

### PR DESCRIPTION
**Pull request by likholat**

### CODE REVIEW

**Comment by dkurt** [comment]
please choose `develop` as target branch

**Comment by dkurt** [comment]
Please also remove a file with classes names

File: README.md [hunk 1]
```diff
@@ -22,6 +22,20 @@ Top 5 accuracy: 0.89814
 
 # openvino_quantization
 
+Download ```convert.py``` script:
+https://gist.github.com/dkurt/3486f09fafe568ee20e500310dedcb9f
```
Comment (dkurt) [inline README.md]: Please just insert a content of this script here

File: README.md [hunk 2]
```diff
@@ -22,6 +22,20 @@ Top 5 accuracy: 0.89814
...
+
+Run the convert.py before converting the model, this script creates resnet_v2_101_299_opt.pb file in current directory:
+``bash
+python convert.py --graph path/to/resnet_v2_101_299_frozen.pb
+``
+
+To convert TensorFlow model to Intermediate Representation:
+``bash
+cd folder/for/IR/model
+python path/to/openvino/deployment_tools/model_optimizer/mo_tf.py --input_model path/to/resnet_v2_101_299_opt.pb --input_shape "[1,299,299,3]"
```
Comment (dkurt) [inline README.md]: Just `/opt/openvino/deployment_tools/model_optimizer/mo_tf.py`

File: README.md (hunk 1 above)
Comment (dkurt) [reply README.md]: Can you please check if master branch of OpenVINO will work or not?
```
git clone https://github.com/openvinotoolkit/openvino
python3 ~/openvino/model-optimizer/mo_tf.py ...
```

File: README.md [hunk 3]
```diff
@@ -22,9 +22,46 @@ Top 5 accuracy: 0.89814
...
+        graph_def.ParseFromString(f.read())
+except:
+    with tf.gfile.FastGFile(pb_file, 'rb') as f:
+        graph_def.ParseFromString(f.read())
+
+graph_def = optimize_for_inference_lib.optimize_for_inference(graph_def, ['input'], ['output'], tf.float32.as_datatype_enum)
+
+with tf.gfile.FastGFile('resnet_v2_101_299_opt.pb', 'wb') as f:
+   f.write(graph_def.SerializeToString())t
+``
```
Comment (dkurt) [inline README.md]: Please check rendered README.md - it's unreadable https://github.com/likholat/openVINO_quantization/blob/openvino_test/README.md

File: classification/classification_opvn.py [hunk 4]
```diff
@@ -0,0 +1,39 @@
+import tensorflow as tf
+import numpy as np
+import cv2 as cv
```
Comment (dkurt) [inline classification/classification_opvn.py:3]: There is no much difference between `classification_tf` and `classification_opvn` - please create a single `classification.py` which accepts flag for target engine

File: demo_classification_opvn.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
+import numpy as np
+import cv2 as cv
+import argparse
+import os
```
Comment (dkurt) [inline demo_classification_opvn.py]: There is no much difference between `demo_classification_tf` and `demo_classification_opvn` - please create a single `demo.py` which accepts flag for target engine.

The same for `evaluate.py`

File: README.md [hunk 6]
```diff
@@ -22,9 +22,46 @@ Top 5 accuracy: 0.89814
...
+To convert TensorFlow model to Intermediate Representation:
+``bash
+cd folder/for/IR/model
+python ./opt/openvino/deployment_tools/model_optimizer/mo_tf.py --input_model path/to/resnet_v2_101_299_opt.pb --input_shape "[1,299,299,3]"
+``
+
 To initialise OpenVINO environment variables open the Command Prompt, and run the setupvars.bat batch file:
 ``bash
-cd C:\Program Files (x86)\IntelSWTools\openvino\bin\
+cd C:/Program Files (x86)/IntelSWTools/openvino/bin/
```
Comment (dkurt) [inline README.md]: Please don't mix Linux and Windows commands. Choose one

File: README.md [hunk 7]
```diff
@@ -22,9 +22,46 @@ Top 5 accuracy: 0.89814
...
+
+Run the script before converting the model, it creates resnet_v2_101_299_opt.pb file in current directory:
+``bash
+python convert.py --graph path/to/resnet_v2_101_299_frozen.pb
+``
+
+To convert TensorFlow model to Intermediate Representation:
+``bash
+cd folder/for/IR/model
+python ./opt/openvino/deployment_tools/model_optimizer/mo_tf.py --input_model path/to/resnet_v2_101_299_opt.pb --input_shape "[1,299,299,3]"
```
Comment (dkurt) [inline README.md]: replace to `--input_model resnet_v2_101_299_opt.pb`

File: README.md [hunk 8]
```diff
@@ -22,9 +22,46 @@ Top 5 accuracy: 0.89814
...
+
+graph_def = optimize_for_inference_lib.optimize_for_inference(graph_def, ['input'], ['output'], tf.float32.as_datatype_enum)
+
+with tf.gfile.FastGFile('resnet_v2_101_299_opt.pb', 'wb') as f:
+   f.write(graph_def.SerializeToString())t
+``
+
+Run the script before converting the model, it creates resnet_v2_101_299_opt.pb file in current directory:
+``bash
+python convert.py --graph path/to/resnet_v2_101_299_frozen.pb
```
Comment (dkurt) [inline README.md]: `--graph resnet_v2_101_299_frozen.pb`

File: README.md (hunk 3 above)
Comment (dkurt) [reply README.md]: It's even better to move this script to root folder

File: calibration.py [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    def value(self):
+        """ Returns accuracy metric value for the last model output. """
+        return {self.name: [self._values[-1]]}
+
+    @property
+    def avg_value(self):
+        """ Returns accuracy metric value for all model outputs. """
+        print(len(self._values))    #print res: 300
+        print("Round #{}  {} = {}".format(self.round, self.name, sum(self._values)/len(self._values)))
+        self.round += 1
```
Comment (dkurt) [inline calibration.py]: What is `round`?

File: calibration.py [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    def update(self, outputs, labels):
+        """ Updates prediction matches.
+        :param output: model output
+        :param target: annotations
+
+        Put your post-processing code here.
+        Put your custom metric code here.
+        The metric gets appended to the list of metric values
+        """
+        result = np.argsort(outputs[0][0])[::-1][:5]
```
Comment (dkurt) [inline calibration.py]: Do we need top five elements? If not, just do `result = np.argsort(outputs.reshape(-1))[-1]`

File: calibration.py [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+        if (result[0] == labels[0]):
+            self._values.append(1)
+        else:
+            self._values.append(0)
```
Comment (dkurt) [inline calibration.py]:
```python
self._values.append(1 if result[0] == labels[0] else 0)
```

File: calibration.py [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+        """ Resets collected matches """
+        self._values = []
+
+    @property
+    def higher_better(self):
+        """Attribute whether the metric should be increased"""
+        return True
+
+    def get_attributes(self):
+        return {self.name: {"direction": "higher-better", "type": ""}}
```
Comment (dkurt) [inline calibration.py]: Will script work without `get_attributes` and `higher_better`?

File: calibration.py [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+            'preset': 'performance',
+            'stat_subset_size': 300
+        }
+    }
+]
+
+model = load_model(model_config)
+
+data_loader = MyDataLoader(dataset_config)
+metric = MyMetric()
```
Comment (dkurt) [inline calibration.py]: Please name metric and DataLoader correspondingly to what they work with

File: demo_classification.py [hunk 14]
```diff
@@ -2,26 +2,38 @@
 import numpy as np
 import cv2 as cv
 import argparse
+import os
+
 from classification.classification_tf import TensorFlowClassification
+from classification.classification_opvn import OpenVinoClassification
```
Comment (dkurt) [inline demo_classification.py]: Please import corresponding module depends on `--engine`:
```python
if args.engine == 'tf':
    from classification.classification_tf import TensorFlowClassification
    engine = TensorFlowClassification()
else:
    from classification.classification_opvn import OpenVinoClassification
    engine = OpenVinoClassification()

File: demo_classification.py [hunk 15]
```diff
@@ -2,26 +2,38 @@
...
     with open(argv.labels, 'rt') as f:
-        labels = f.read().strip().split('\n') 
+        labels = f.read().strip().split('\n')
+
+    img = cv.imread(argv.image)  
 
-    img = cv.imread(argv.image) 
+    if (argv.engine == 'tf'):
+        network = TensorFlowClassification(argv.graph)
+        probs, indices = network.classify(img, top_k = 5)
```
Comment (dkurt) [inline demo_classification.py]: Why it's under `if`?

File: evaluate.py [hunk 16]
```diff
@@ -2,23 +2,33 @@
...
 
     top_1 = 0
     top_5 = 0
 
+    if (argv.engine == 'tf'):
+        network = TensorFlowClassification(argv.graph)
+        
+    elif (argv.engine == 'opvn'):
+        model_xml = os.path.join(argv.graph, 'resnet_v2_101_299_opt.xml')
+        model_bin = os.path.join(argv.graph, 'resnet_v2_101_299_opt.bin')
```
Comment (dkurt) [inline evaluate.py]: What is `args.graph` in this way?

File: classification/classification_opvn.py [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+        image_size_x = 299
+        image_size_y = 299 
+
+        img = cv.resize(img, (image_size_x, image_size_y))
+        img = np.expand_dims(img, axis=0)
+        img = (img - image_mean) / image_std
+        img = img.astype(np.float32)
+        img = img.transpose((0, 3, 1, 2))
+
+        exec_net = self.ie.load_network(network=self.net, device_name='CPU')
```
Comment (dkurt) [inline classification/classification_opvn.py]: should be inside `__init__`

File: README.md [hunk 18]
```diff
@@ -7,33 +7,59 @@ https://storage.googleapis.com/download.tensorflow.org/models/tflite_11_05_08/re
 
 To validate TensorFlow model on one image run:
 ```bash
-python demo_classification_tf.py --graph /path/to/resnet_v2_101_299_frozen.pb --image /path/to/example.jpeg
+python3 demo_classification.py --graph /path/to/resnet_v2_101_299_frozen.pb --image /path/to/example.jpeg --engine tf
```
Comment (dkurt) [inline README.md]: please avoid extra information in the example command line. It's hard for users to modify it. Just
```
python3 demo_classification.py --graph resnet_v2_101_299_frozen.pb --engine tf --image example.jpeg
```

File: README.md [hunk 19]
```diff
@@ -7,33 +7,59 @@ https://storage.googleapis.com/download.tensorflow.org/models/tflite_11_05_08/re
...
 ```
 
 To validate TensorFlow model on ImageNet dataset run:
 ```bash
-python3 evaluate_tf.py --graph /path/to/resnet_v2_101_299_frozen.pb --dataset /path/to/ILSVRC2012_img_val
+python3 evaluate.py --graph /path/to/resnet_v2_101_299_frozen.pb --dataset /path/to/ILSVRC2012_img_val --engine tf
```
Comment (dkurt) [inline README.md]: the same as above

File: README.md [hunk 20]
```diff
@@ -7,33 +7,59 @@ https://storage.googleapis.com/download.tensorflow.org/models/tflite_11_05_08/re
...
 ```
 Estimated accuracy:
 ```bash
 Top 1 accuracy: 0.69054
 Top 5 accuracy: 0.89814
 ```
 
-# openvino_quantization
+# openvino_sample
```
Comment (dkurt) [inline README.md]: Please do headers more readable
```
## OpenVINO sample
```

File: README.md [hunk 21]
```diff
@@ -7,33 +7,59 @@ https://storage.googleapis.com/download.tensorflow.org/models/tflite_11_05_08/re
...
+Run the ```convert.py``` script before converting the model, it creates resnet_v2_101_299_opt.pb file in current directory:
 ``bash
-cd C:\Program Files (x86)\IntelSWTools\openvino\bin\
-setupvars.bat
+python3 convert.py --graph resnet_v2_101_299_frozen.pb
+``
+
+To convert TensorFlow model to Intermediate Representation:
+``bash
+cd folder/for/IR/model
```
Comment (dkurt) [inline README.md]: Not needed because user can specify full path by arguments

File: README.md [hunk 22]
```diff
@@ -7,33 +7,59 @@ https://storage.googleapis.com/download.tensorflow.org/models/tflite_11_05_08/re
...
 
 To validate OpenVINO model on ImageNet dataset run:
 ``bash
-python evaluate_opvn.py --graph /path/to/resnet_converted --dataset /path/to/ILSVRC2012_img_val
+python3 evaluate.py --graph /path/to/resnet_converted --dataset ILSVRC2012_img_val --engine opvn
+``
+
+To quantizes the model to int8 model run:
+``bash
+python3 calibration.py --xml resnet_v2_101_299_opt.xml --bin resnet_v2_101_299_opt.bin --data ILSVRC2012_img_val --annotation ILSVRC2012_img_val/val.txt
```
Comment (dkurt) [inline README.md]: Please let user specify only path to `.xml`

File: evaluate.py (hunk 16 above)
Comment (likholat) [reply evaluate.py]: Path to folder with model

File: calibration.py (hunk 10 above)
Comment (likholat) [reply calibration.py]: ```AttributeError: 'list' object has no attribute 'reshape'```

File: README.md [hunk 23]
```diff
@@ -7,33 +7,83 @@ https://storage.googleapis.com/download.tensorflow.org/models/tflite_11_05_08/re
...
+Estimated accuracy for OpenVINO model:
+```bash
+Top 1 accuracy: 0.69054
+Top 5 accuracy: 0.89814
+```
+Estimated accuracy for int8 model:
+```bash
+Top 1 accuracy: 0.69316
+Top 5 accuracy: 0.89776
+```
```
Comment (dkurt) [inline README.md:72]: please format as table

File: README.md (hunk 1 above)
Comment (likholat) [reply README.md]: No, it doesn't work

File: README.md (hunk 21 above)
Comment (likholat) [reply README.md]: ```[ ERROR ]  Both --input_model and --saved_model_dir are defined. Specify either input model or saved model directory.```

Comment (dkurt) [reply README.md]: I mean we don't need to `cd` to the folder with model - user can specify the path by `--input_model`

**Pull request merged**